### ISSUE TITLE
Security of unencrypted git:// protocol

### ISSUE THREAD

**Issue by nullchinchilla** [JuliaLang/METADATA.jl#7495]
It appears that most of the packages are registered with the unencrypted, unauthenticated `git://` protocol rather than the HTTPS protocol that GitHub etc also supports.

This doesn't seem to be a very good idea. Plaintext Git can easily be tampered with, and essentially installing packages right now involves downloading arbitrary code from an untrustworthy network. There is no way to ensure that the code you download is actually the real code from the repo, and not malware from somebody who controls your WiFi etc.

A system like [The Update Framework](https://theupdateframework.github.io/) would be ideal, but the lowest-hanging fruit is probably to simply serve everything over Git over HTTPS, rather than plain Git.

**Comment by tkelman** [comment]
ref https://github.com/JuliaLang/METADATA.jl/issues/400, https://github.com/JuliaLang/METADATA.jl/pull/402, and https://github.com/JuliaLang/METADATA.jl/pull/3752 for some history. We currently allow either, with more recently-registered packages now defaulting to https. We probably should modify the rest en masse since I think there haven't been too many issues overall with more recently-registered packages.

**Comment by tkelman** [comment]
And there's also an option in the package manager, either via gitconfig for Julia 0.4 or `Pkg.setprotocol!` in 0.5-and-newer to always rewrite git:// to https:// if you so choose. I think we might even be normalizing github urls to https by default already.

**Comment by bunsim** [comment]
Is METADATA itself accessed using https or git?

**Comment by tkelman** [comment]
https since 0.5 https://github.com/JuliaLang/julia/blob/v0.5.0/base/pkg/pkg.jl#L9

**Comment by bunsim** [comment]
Ah I see. I guess I'll install the latest Julia then; I'm using Debian unstable's repos which doesn't have 0.5 yet.

**Comment by tkelman** [comment]
Yeah there are some issues with debian packaging preventing 0.5 from being packaged. You can check the discussions on the debian pkg-julia-devel list. I don't think they've yet got a version of libgit2 that can use https packaged in debian, due to openssl licensing issues. https://github.com/libgit2/libgit2/pull/3935 might unblock that another way, if it ever gets finished and merged.

**Comment by bunsim** [comment]
Ideally, though, there should be a signing system independent of trusting GitHub's HTTPS. Package management is a place where security is extremely important, since otherwise any machine with Julia on it would essentially have a huge gaping backdoor. Even if it's all HTTPS, the current system means that anybody who hacks GitHub (or any GitHub account with write access to METADATA) suddenly has unlimited control over everybody using Julia packages.

Perhaps integrating Git's commit or tag signing support would be possible? The METADATA repository could be signed, and one could require individual packages to be signed with a key registered with the METADATA repo; this would obviate the need for HTTPS, and would also make mirrored repos equally trustworthy.

**Comment by tkelman** [comment]
Can commits and tags be signed via libgit2? Does that impose a gpg dependency?

**Comment by bunsim** [comment]
Ah yeah that'll probably impose a GPG dependency :( I forgot that Julia used libgit2 rather than calling `git` like many other package managers.

**Comment by tkelman** [comment]
What does rust do? They use libgit2 as well in cargo IIRC.

**Comment by bunsim** [comment]
They currently use a central trusted metadata server system (like julia, except entirely over HTTPS etc), similar to Node's `npm` etc. However, they appear to be seriously considering moving to a TUF-based system: https://github.com/rust-lang/crates.io/issues/75

**Comment by DilumAluthge** [comment]
METADATA.jl is no longer maintained, so I am closing this issue.

If this issue is still relevant, please re-open a new issue on the Julia General registry: https://github.com/JuliaRegistries/General
